# Домашнее задание на проверку знаний библиотеки для анализа данных Pandas.

Вам представлен датасет, который содержит информацию по результатам сезона 2016 года игры в бейсбол.

Вам необходимо загрузить датасет (он находится в архиве с текущим домашним заданием). Провести его очистку, если это необходимо и ответить на поставленные аналитические
вопросы.

Список колонок, представленных в датасете:

---
* attendance        - количество зрителей матча
* away_team         - название гостевой команды
* away_team_errors  - количество ошибок гостевой команды
* away_team_hits    - количество хитов гостевой команды (хит в бейсболе - удар при котором отбивающий достиг первой базы)
* away_team_runs    - количество ранов гостевой команды (ран в бейсболе - очко заработанное игроком нападения)
* date              - дата проведения матча
* field_type        - тип игрового поля
* game_type         - тип игры
* home_team         - название домашней команды
* home_team_errors  - количество ошибок домашней команды
* home_team_hits    - количество хитов домашней команды (хит в бейсболе - удар при котором отбивающий достиг первой базы)
* home_team_runs    - количество ранов домашней команды (ран в бейсболе - очко заработанное игроком нападения)
* start_time        - время начала игрового матча
* venue             - название места проведения матча (стадион, поле, арена)
* day_of_week       - день недели проведения матча
* temperature       - температура воздуха в день проведения матча в фаренгейтах
* wind_speed        - скорость ветра в день проведения матча
* wind_direction    - направление скорости ветра
* sky               - облачность
* total_runs        - общее количество ранов для двух команд
* game_hours_dec    - длительность матча указаная в часах
* season            - тип игрового сезона
* home_team_win     - результат домашней команды (1 - выигрыш)
* home_team_loss    - результат домашней команды (0 - проигрыш)
* home_team_outcome - исход матча
---

Всего в задании 20 вопросов. За каждый правильный вопрос начисляется 5 баллов. Таким образом за все правильные ответы вы получите - 100 баллов. Оценка затем будет приведена к 10-ти бальной шкале.

---

**Кроме того, баллы могут быть снижены за**:

- 1 балл снижается за качество кода - старайтесь чтобы код был максимально лаконичным. Названия переменных (если они используются) должны отражать максимальную ее сущность. Вывод старайтесь делать минимальным, так как первичная цель - сделать качественный анализ, чтобы быстро ответить на поставленный вопрос (никаких таблиц на 100 строк).


- 5 баллов снижается если ответ на домашнюю работу будет прислан в виде скриншота, ноутбука, текстового файла и прочих способов. Используйте при отправке на проверку только GitHub/GitLab либо прочий сервис. Присылать домашнюю работу необходимо ссылкой на ваш репозиторий на адрес электронной почты maratmovlamov2017@gmail.com


- за каждый день просрочки дедлайна снижается по 10 баллов. Один день - минус 10 баллов, второй день минус 10 баллов, третий день минус 10 баллов. Соблюдайте сроки и не приступайте к домашнему заданию в последний момент.

---

Удачи вам в поисках истинны :)

In [2]:
# Импортируйте все необходимые библиотеки и модули которые вам необходимы
import pandas as pd
import numpy as np

In [3]:
# Загрузите датасет и проведите его очистку и форматирование если это необходимо
data = pd.read_csv('./baseball_games.csv')
data.describe()

Unnamed: 0    attendance  away_team_errors  away_team_hits  \
count  2463.000000   2460.000000       2463.000000     2463.000000   
mean   1231.000000  30370.703659          0.580593        8.764515   
std     711.151179   9875.466703          0.793391        3.511581   
min       0.000000   8766.000000          0.000000        1.000000   
25%     615.500000  22432.000000          0.000000        6.000000   
50%    1231.000000  30604.500000          0.000000        8.000000   
75%    1846.500000  38396.250000          1.000000       11.000000   
max    2462.000000  54449.000000          5.000000       22.000000   

       away_team_runs  home_team_errors  home_team_hits  home_team_runs  \
count     2463.000000       2463.000000     2463.000000     2463.000000   
mean         4.413723          0.585871        8.611855        4.519691   
std          3.104556          0.805542        3.436965        3.111572   
min          0.000000          0.000000        0.000000        0.000000   
25%          2.000000          0.000000        6.000000        2.000000   
50%          4.000000          0.000000        8.000000        4.000000   
75%          6.000000          1.000000       11.000000        6.000000   
max         21.000000          5.000000       22.000000       17.000000   

       temperature   wind_speed   total_runs  game_hours_dec  home_team_win  \
count  2463.000000  2463.000000  2463.000000     2463.000000    2463.000000   
mean     73.818920     7.494925     8.933415        3.085302       0.529436   
std      10.583106     5.055970     4.480352        0.460491       0.499234   
min      31.000000     0.000000     1.000000        1.250000       0.000000   
25%      67.000000     4.000000     6.000000        2.783333       0.000000   
50%      74.000000     7.000000     8.000000        3.033333       1.000000   
75%      81.000000    11.000000    12.000000        3.308333       1.000000   
max     101.000000    25.000000    29.000000        6.216667       1.000000   

       home_team_loss  
count     2463.000000  
mean         0.470158  
std          0.499210  
min          0.000000  
25%          0.000000  
50%          0.000000  
75%          1.000000  
max          1.000000

 Вижу непонятный столбик 'Unnamed: 0', проверяю содержимое, и не найдя там ничего полезного удаляю

In [5]:
data.drop(columns=['Unnamed: 0'], axis=1, inplace=True)

Просматриваю список колонок и сокращаю некоторые названия

In [7]:
data.columns

Index(['attendance', 'away_team', 'away_team_errors', 'away_team_hits',
       'away_team_runs', 'date', 'field_type', 'game_type', 'home_team',
       'home_team_errors', 'home_team_hits', 'home_team_runs', 'start_time',
       'venue', 'day_of_week', 'temperature', 'wind_speed', 'wind_direction',
       'sky', 'total_runs', 'game_hours_dec', 'season', 'home_team_win',
       'home_team_loss', 'home_team_outcome'],
      dtype='object')

In [8]:
data.columns = ['attendance', 'away_team', 'a_t_errors', 'a_t_hits',
               'a_t_runs', 'date', 'field_type', 'game_type', 'home_team',
               'h_t_errors', 'h_t_hits', 'h_t_runs', 'start_time',
               'venue', 'day_of_week', 'temp', 'wind_speed', 'wind_direction',
               'sky', 'total_runs', 'game_hours_dec', 'season', 'h_t_win',
               'h_t_loss', 'h_t_outcome']

Вижу что отображаются не все колонки, по этому задаю отображение всей таблицы

In [10]:
pd.set_option('display.max_columns', None)
data.head(10)

attendance              away_team  a_t_errors  a_t_hits  a_t_runs  \
0     40030.0          New York Mets           1         7         3   
1     21621.0  Philadelphia Phillies           0         5         2   
2     12622.0        Minnesota Twins           0         5         2   
3     18531.0   Washington Nationals           0         8         3   
4     18572.0       Colorado Rockies           1         8         4   
5     28386.0       Seattle Mariners           1        11        10   
6     12757.0      Toronto Blue Jays           0         9         2   
7     28329.0    Los Angeles Dodgers           0         6         3   
8     26049.0    St. Louis Cardinals           1         8         5   
9     10478.0      Chicago White Sox           0        11         5   

         date field_type   game_type             home_team  h_t_errors  \
0  2016-04-03   on grass  Night Game    Kansas City Royals           0   
1  2016-04-06   on grass  Night Game       Cincinnati Reds           0   
2  2016-04-06   on grass  Night Game     Baltimore Orioles           0   
3  2016-04-06   on grass  Night Game        Atlanta Braves           1   
4  2016-04-06   on grass    Day Game  Arizona Diamondbacks           0   
5  2016-04-05   on grass  Night Game         Texas Rangers           1   
6  2016-04-05    on turf  Night Game        Tampa Bay Rays           1   
7  2016-04-05   on grass  Night Game      San Diego Padres           1   
8  2016-04-05   on grass  Night Game    Pittsburgh Pirates           2   
9  2016-04-05   on grass  Night Game     Oakland Athletics           0   

   h_t_hits  h_t_runs        start_time                            venue  \
0         9         4   7:38 p.m. Local                 Kauffman Stadium   
1         8         3   7:11 p.m. Local         Great American Ball Park   
2         9         4   7:07 p.m. Local      Oriole Park at Camden Yards   
3         8         1   7:10 p.m. Local                     Turner Field   
4         8         3  12:40 p.m. Local                      Chase Field   
5         7         2   7:07 p.m. Local     Globe Life Park in Arlington   
6         7         3   7:10 p.m. Local                  Tropicana Field   
7         2         0   7:11 p.m. Local                       Petco Park   
8        12         6   7:08 p.m. Local                         PNC Park   
9        10         4   7:08 p.m. Local  Oakland-Alameda County Coliseum   

  day_of_week  temp  wind_speed         wind_direction       sky  total_runs  \
0      Sunday  74.0        14.0     from Right to Left     Sunny           7   
1   Wednesday  55.0        24.0     from Right to Left  Overcast           5   
2   Wednesday  48.0         7.0       out to Leftfield   Unknown           6   
3   Wednesday  65.0        10.0     from Right to Left    Cloudy           4   
4   Wednesday  77.0         0.0   in unknown direction   In Dome           7   
5     Tuesday  81.0        12.0     from Right to Left    Cloudy          12   
6     Tuesday  72.0         0.0   in unknown direction   In Dome           5   
7     Tuesday  70.0         6.0     from Left to Right   Unknown           3   
8     Tuesday  40.0         7.0       out to Leftfield     Sunny          11   
9     Tuesday  70.0         8.0      out to Rightfield    Cloudy           9   

   game_hours_dec          season  h_t_win  h_t_loss h_t_outcome  
0        3.216667  regular season        1         0         Win  
1        2.383333  regular season        1         0         Win  
2        3.183333  regular season        1         0         Win  
3        2.883333  regular season        0         1        Loss  
4        2.650000  regular season        0         1        Loss  
5        3.500000  regular season        0         1        Loss  
6        3.116667  regular season        1         0         Win  
7        2.600000  regular season        0         1        Loss  
8        3.450000  regular season        1         0         Win  
9      

Вижу, что одна колонка "attendance" имеет пустые значения, из 2463 строки 3 пустые, заполняем медианным значением

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2463 entries, 0 to 2462
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   attendance      2460 non-null   float64
 1   away_team       2463 non-null   object 
 2   a_t_errors      2463 non-null   int64  
 3   a_t_hits        2463 non-null   int64  
 4   a_t_runs        2463 non-null   int64  
 5   date            2463 non-null   object 
 6   field_type      2463 non-null   object 
 7   game_type       2463 non-null   object 
 8   home_team       2463 non-null   object 
 9   h_t_errors      2463 non-null   int64  
 10  h_t_hits        2463 non-null   int64  
 11  h_t_runs        2463 non-null   int64  
 12  start_time      2463 non-null   object 
 13  venue           2463 non-null   object 
 14  day_of_week     2463 non-null   object 
 15  temp            2463 non-null   float64
 16  wind_speed      2463 non-null   float64
 17  wind_direction  2463 non-null   o

In [13]:
data.loc[pd.isnull( data['attendance'])].head()

attendance            away_team  a_t_errors  a_t_hits  a_t_runs  \
220          NaN  St. Louis Cardinals           1         8         3   
1724         NaN        New York Mets           0         5         1   
1912         NaN    Cleveland Indians           2         7         6   

            date field_type game_type           home_team  h_t_errors  \
220   2016-07-26   on grass  Day Game       New York Mets           1   
1724  2016-06-07   on grass  Day Game  Pittsburgh Pirates           0   
1912  2016-05-23   on grass  Day Game   Chicago White Sox           0   

      h_t_hits  h_t_runs       start_time                venue day_of_week  \
220          7         2  4:11 p.m. Local           Citi Field     Tuesday   
1724        10         3  4:08 p.m. Local             PNC Park     Tuesday   
1912        10         7  4:10 p.m. Local  U.S. Cellular Field      Monday   

      temp  wind_speed       wind_direction     sky  total_runs  \
220   90.0        11.0    out to Rightfield  Cloudy           5   
1724  65.0        17.0   out to Centerfield  Cloudy           4   
1912  82.0        11.0    in from Leftfield   Sunny          13   

      game_hours_dec          season  h_t_win  h_t_loss h_t_outcome  
220         3.300000  regular season        0         1        Loss  
1724        2.666667  regular season        1         0         Win  
1912        3.166667  regular season        1         0         Win

In [14]:
median_attendance = round(data['attendance'].median(), 0)
data.fillna( {'attendance': median_attendance}, inplace = True )
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2463 entries, 0 to 2462
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   attendance      2463 non-null   float64
 1   away_team       2463 non-null   object 
 2   a_t_errors      2463 non-null   int64  
 3   a_t_hits        2463 non-null   int64  
 4   a_t_runs        2463 non-null   int64  
 5   date            2463 non-null   object 
 6   field_type      2463 non-null   object 
 7   game_type       2463 non-null   object 
 8   home_team       2463 non-null   object 
 9   h_t_errors      2463 non-null   int64  
 10  h_t_hits        2463 non-null   int64  
 11  h_t_runs        2463 non-null   int64  
 12  start_time      2463 non-null   object 
 13  venue           2463 non-null   object 
 14  day_of_week     2463 non-null   object 
 15  temp            2463 non-null   float64
 16  wind_speed      2463 non-null   float64
 17  wind_direction  2463 non-null   o

Увидел в задании отбор по температуре, решил перевести Фарингейты в более понятные нам Цельсии с округлением до одной десятой. И переименовал колонку.

In [16]:
data['temp'] = data['temp'].apply(lambda x: round((x-32) * 5/9, 1))
data.rename( columns={'temp': 'temp_C'}, inplace=True)
data

attendance              away_team  a_t_errors  a_t_hits  a_t_runs  \
0        40030.0          New York Mets           1         7         3   
1        21621.0  Philadelphia Phillies           0         5         2   
2        12622.0        Minnesota Twins           0         5         2   
3        18531.0   Washington Nationals           0         8         3   
4        18572.0       Colorado Rockies           1         8         4   
...          ...                    ...         ...       ...       ...   
2458     31042.0      Toronto Blue Jays           2         7         5   
2459     39500.0    St. Louis Cardinals           0         5         1   
2460     20098.0   San Francisco Giants           0         6         3   
2461     17883.0         Detroit Tigers           0        13         7   
2462     10298.0         Boston Red Sox           1        10         6   

            date field_type   game_type             home_team  h_t_errors  \
0     2016-04-03   on grass  Night Game    Kansas City Royals           0   
1     2016-04-06   on grass  Night Game       Cincinnati Reds           0   
2     2016-04-06   on grass  Night Game     Baltimore Orioles           0   
3     2016-04-06   on grass  Night Game        Atlanta Braves           1   
4     2016-04-06   on grass    Day Game  Arizona Diamondbacks           0   
...          ...        ...         ...                   ...         ...   
2458  2016-04-03    on turf    Day Game        Tampa Bay Rays           1   
2459  2016-04-03   on grass    Day Game    Pittsburgh Pirates           1   
2460  2016-04-06   on grass    Day Game     Milwaukee Brewers           2   
2461  2016-04-06   on grass    Day Game         Miami Marlins           1   
2462  2016-04-06   on grass  Night Game     Cleveland Indians           0   

      h_t_hits  h_t_runs        start_time                        venue  \
0            9         4   7:38 p.m. Local             Kauffman Stadium   
1            8         3   7:11 p.m. Local     Great American Ball Park   
2            9         4   7:07 p.m. Local  Oriole Park at Camden Yards   
3            8         1   7:10 p.m. Local                 Turner Field   
4            8         3  12:40 p.m. Local                  Chase Field   
...        ...       ...               ...                          ...   
2458         7         3   4:09 p.m. Local              Tropicana Field   
2459         9         4   1:15 p.m. Local                     PNC Park   
2460         9         4  12:41 p.m. Local                  Miller Park   
2461        10         3   4:57 p.m. Local                 Marlins Park   
2462         9         7   6:22 p.m. Local            Progressive Field   

     day_of_week  temp_C  wind_speed         wind_direction       sky  \
0         Sunday    23.3        14.0     from Right to Left     Sunny   
1      Wednesday    12.8        24.0     from Right to Left  Overcast   
2      Wednesday     8.9         7.0       out to Leftfield   Unknown   
3      Wednesday    18.3        10.0     from Right to Left    Cloudy   
4      Wednesday    25.0         0.0   in unknown direction   In Dome   
...          ...     ...         ...                    ...       ...   
2458      Sunday    22.2         0.0   in unknown direction   In Dome   
2459      Sunday     3.9        14.0       out to Leftfield   Unknown   
2460   Wednesday    18.9         0.0   in unknown direction   In Dome   
2461   Wednesday    21.7         0.0   in unknown direction   In Dome   
2462   Wednesday    15.6         7.0       out to Leftfield   Unknown   

      total_runs  game_hours_dec          season  h_t_win  h_t_loss  \
0              7        3.216667  regular season        1         0   
1              5        2.383333  regular season        1         0   
2              6        3.183333  regular season        1         0   
3              4        2.883333  regular season        0         1   
4              7        2.650000  regular seaso

Очистка и форматирование завершены

# 1. На какую игру пришло максимальное количество зрителей за весь сезон игр?

Отсортируем по количеству зрителей, при помощи sort_values и выведем первый результат, при помощи head(1), ascending=False - означает сортировка по убыванию. Максимальное количество зрителей 54_449 человек.

In [20]:
data.sort_values( by='attendance', ascending=False).head(1)

attendance     away_team  a_t_errors  a_t_hits  a_t_runs        date  \
360     54449.0  Chicago Cubs           2        13        10  2016-10-19   

    field_type   game_type            home_team  h_t_errors  h_t_hits  \
360   on grass  Night Game  Los Angeles Dodgers           4         6   

     h_t_runs       start_time           venue day_of_week  temp_C  \
360         2  5:10 p.m. Local  Dodger Stadium   Wednesday    31.7   

     wind_speed      wind_direction    sky  total_runs  game_hours_dec  \
360         4.0   out to Rightfield  Night          12        3.966667   

          season  h_t_win  h_t_loss h_t_outcome  
360  post season        0         1        Loss

# 2. Какая игра была самая холодная (temperature) за весь сезон?

Сортируем по колонке температуры и выводим первую строку при помощи head(1). Температура самой холодной игры сезона -0.6 градуса

In [23]:
data.sort_values( by='temp_C' ).head(1)

attendance         away_team  a_t_errors  a_t_hits  a_t_runs  \
2409     32419.0  New York Yankees           1        13         8   

            date field_type game_type       home_team  h_t_errors  h_t_hits  \
2409  2016-04-09   on grass  Day Game  Detroit Tigers           1         7   

      h_t_runs       start_time          venue day_of_week  temp_C  \
2409         4  1:10 p.m. Local  Comerica Park    Saturday    -0.6   

      wind_speed       wind_direction     sky  total_runs  game_hours_dec  \
2409        18.0   from Left to Right  Cloudy          12        3.333333   

              season  h_t_win  h_t_loss h_t_outcome  
2409  regular season        0         1        Loss

# 3. Какая игра была самая теплая за весь сезон?

Так же сортируем, но выводим первую строку с конца tail(1), температура самой тёплой игры сезона 38.3 градуса

In [26]:
data.sort_values( by='temp_C' ).tail(1)

attendance             away_team  a_t_errors  a_t_hits  a_t_runs  \
2026     21753.0  San Francisco Giants           0         8         3   

            date field_type   game_type             home_team  h_t_errors  \
2026  2016-05-13   on grass  Night Game  Arizona Diamondbacks           0   

      h_t_hits  h_t_runs       start_time        venue day_of_week  temp_C  \
2026         8         1  6:40 p.m. Local  Chase Field      Friday    38.3   

      wind_speed         wind_direction    sky  total_runs  game_hours_dec  \
2026         9.0   in unknown direction  Sunny           4             3.0   

              season  h_t_win  h_t_loss h_t_outcome  
2026  regular season        0         1        Loss

# 4. Какая игра в сезоне была самая долгая по продолжительности матча?

Сортируем сразу по двум колонкам, сперва указываем основной сезон, а затем сортируем по продолжительности. В данном случае используем атрибут ascending=[False, False], что означает сортировка по убыванию каждой из колонок

In [29]:
most_long_game = data.sort_values( by=['season', 'game_hours_dec'], ascending=[False, False]).head(1)
most_long_game

attendance          away_team  a_t_errors  a_t_hits  a_t_runs  \
1445     45825.0  Cleveland Indians           0        15         2   

            date field_type game_type          home_team  h_t_errors  \
1445  2016-07-01    on turf  Day Game  Toronto Blue Jays           2   

      h_t_hits  h_t_runs       start_time          venue day_of_week  temp_C  \
1445        11         1  1:20 p.m. Local  Rogers Centre      Friday    20.0   

      wind_speed         wind_direction      sky  total_runs  game_hours_dec  \
1445         0.0   in unknown direction  In Dome           3        6.216667   

              season  h_t_win  h_t_loss h_t_outcome  
1445  regular season        0         1        Loss

Создадим функцию для конвертации десятичного времени в более понятные часы минуты. Функция принимает на вход 2 значения: время в десятичном формате и продолжительность в виде слова, для подстановки в отформатированный текст (долгая или короткая)

In [31]:
def convert(number, duration):
    hours = round(number)
    number -= hours
    minutes = round(number * 60)
    return f'Самая {duration} по продолжительности игра в сезоне длилась {hours} часов {minutes} минут'

print(convert(most_long_game['game_hours_dec'].tolist()[0], 'долгая'))

Самая долгая по продолжительности игра в сезоне длилась 6 часов 13 минут


# 5. Какая игра в сезоне была самая короткая по продолжительности матча?

Аналогично производим сортировку, но для колонки 'game_hours_dec' указываем сортировку True, что означает по возрастанию

In [34]:
least_long_game = data.sort_values( by=['season', 'game_hours_dec'], ascending=[False, True]).head(1)
least_long_game

attendance     away_team  a_t_errors  a_t_hits  a_t_runs        date  \
423     19991.0  Chicago Cubs           1         4         1  2016-09-29   

    field_type   game_type           home_team  h_t_errors  h_t_hits  \
423   on grass  Night Game  Pittsburgh Pirates           0         2   

     h_t_runs       start_time     venue day_of_week  temp_C  wind_speed  \
423         1  7:06 p.m. Local  PNC Park    Thursday    17.2        12.0   

         wind_direction       sky  total_runs  game_hours_dec          season  \
423   in from Leftfield  Overcast           2            1.25  regular season   

     h_t_win  h_t_loss h_t_outcome  
423        0         0        Loss

In [35]:
print(convert(least_long_game['game_hours_dec'].tolist()[0], 'короткая'))

Самая короткая по продолжительности игра в сезоне длилась 1 часов 15 минут


# 6. Сколько матчей в сезоне закончилось ничьей?

Анализируем колонку 'h_t_outcome' с результатами матчей, и приходим к выводу, что всего игр 2463 что равно 1304 + 1159, следовательно ничьёй не закончилась ни одна игра, что подтверждается правилами игры. Т.е. в правилах игры не предусмотрена ничья. "В бейсболе нет ничьи. Она просто не предусмотрена правилами игры. В случае, если к концу девятого иннинга команды набрали равное количество очков, назначаются дополнительные иннинги. И так до победы одной из сторон."

In [38]:
data['h_t_outcome'].value_counts()

h_t_outcome
Win     1304
Loss    1159
Name: count, dtype: int64

# 7. Какая игра была последней в сезоне?

Для выбора последней игры в сезоне я взял 3 метрики: сезон, дату проведения и начало игры по местному времени, соответственно в такой же последовательности и выбиралась сортировка, атрибут ascending для всех 3х сортировок False, то есть по убыванию, соответветственно выводим первую строку в которой будет самая последняя стартовавшая игра в сезоне, по местному времени.

In [41]:
last_game_in_reg_season = data.sort_values( by=['season', 'date', 'start_time'], ascending=[False, False, False]).head(1)
last_game_in_reg_season

attendance          away_team  a_t_errors  a_t_hits  a_t_runs  \
395     36787.0  Toronto Blue Jays           0         9         2   

           date field_type game_type       home_team  h_t_errors  h_t_hits  \
395  2016-10-02   on grass  Day Game  Boston Red Sox           1         4   

     h_t_runs       start_time        venue day_of_week  temp_C  wind_speed  \
395         1  3:17 p.m. Local  Fenway Park      Sunday    13.3         6.0   

        wind_direction      sky  total_runs  game_hours_dec          season  \
395   out to Leftfield  Unknown           3        3.233333  regular season   

     h_t_win  h_t_loss h_t_outcome  
395        0         1        Loss

In [42]:
f"Последняя игра сезона началась {last_game_in_reg_season['date'].to_list()[0]} в {last_game_in_reg_season['start_time'].to_list()[0]}"

'Последняя игра сезона началась 2016-10-02 в 3:17 p.m. Local'

# 8. У какой игры было минимальное количество зрителей?

Минимальное количество зрителей за всё время проевдения игр - 8766

In [45]:
data.sort_values( by='attendance').head(1)

attendance       away_team  a_t_errors  a_t_hits  a_t_runs        date  \
2130      8766.0  Detroit Tigers           0         5         0  2016-05-04   

     field_type   game_type          home_team  h_t_errors  h_t_hits  \
2130   on grass  Night Game  Cleveland Indians           0         5   

      h_t_runs       start_time              venue day_of_week  temp_C  \
2130         4  6:10 p.m. Local  Progressive Field   Wednesday    12.2   

      wind_speed       wind_direction       sky  total_runs  game_hours_dec  \
2130        11.0   from Left to Right  Overcast           4        2.316667   

              season  h_t_win  h_t_loss h_t_outcome  
2130  regular season        1         0         Win

# 9. Какая игра в сезоне была самая ветренная?

Можно было решить опять же через sort_values, но значений с максимальным значением wind_speed было несклоько, что бы вручную не указывать head() воспользовался функцией поиска максимального значения и выдернул это значения по колонке wind_speed. Максимальный показатель ветра был 25 м/с

In [48]:
data[data['wind_speed'] == data['wind_speed'].max()]

attendance          away_team  a_t_errors  a_t_hits  a_t_runs  \
1655     41543.0  Milwaukee Brewers           1        11         5   
2005     35736.0     Houston Astros           2         8         9   

            date field_type   game_type             home_team  h_t_errors  \
1655  2016-06-13   on grass  Night Game  San Francisco Giants           0   
2005  2016-05-15   on grass    Day Game        Boston Red Sox           3   

      h_t_hits  h_t_runs       start_time        venue day_of_week  temp_C  \
1655        14        11  7:15 p.m. Local    AT&T Park      Monday    14.4   
2005        14        10  1:34 p.m. Local  Fenway Park      Sunday    14.4   

      wind_speed       wind_direction     sky  total_runs  game_hours_dec  \
1655        25.0   out to Centerfield  Cloudy          16        3.633333   
2005        25.0    out to Rightfield  Cloudy          19        3.666667   

              season  h_t_win  h_t_loss h_t_outcome  
1655  regular season        1         0         Win  
2005  regular season        1         0         Win

# 10. В какой игре получили максимальное количество очков?

За количество очков отвечает total_runs. Смотрим сколько раз достигались определённые результаты по очкам и видим, что 29 очков (наш максимум) был выбит всего 1 раз, и выводим на экран игру с этим значениемю

In [51]:
data['total_runs'].value_counts().sort_values()

total_runs
29      1
27      1
23      5
24      6
22      9
21     11
20     17
19     31
18     35
2      38
16     42
1      50
17     56
14     74
15     91
12    116
4     118
13    132
10    148
3     157
11    174
6     186
8     195
5     247
7     257
9     266
Name: count, dtype: int64

In [52]:
data[data['total_runs'] == data['total_runs'].max()]

attendance         away_team  a_t_errors  a_t_hits  a_t_runs  \
1788     22588.0  Seattle Mariners           1        16        16   

            date field_type   game_type         home_team  h_t_errors  \
1788  2016-06-02   on grass  Night Game  San Diego Padres           1   

      h_t_hits  h_t_runs       start_time       venue day_of_week  temp_C  \
1788        20        13  6:10 p.m. Local  Petco Park    Thursday    24.4   

      wind_speed      wind_direction    sky  total_runs  game_hours_dec  \
1788        10.0   out to Rightfield  Sunny          29        3.833333   

              season  h_t_win  h_t_loss h_t_outcome  
1788  regular season        0         1        Loss

# 11. Какая игра содержала максимальное количество ошибок домашней команды?

Аналогично с прошлым вопросом смотрим распределение по колонке 'h_t_errors' которое отвечает за количество ошибок домашней команды, сортируем по количеству ошибок и видим, что максимальное количество ошибок сделала домашняя команда лишь единожды. и выводим строку с этой командой по максимальному значению. Максимальное количество ошибок 5.

In [55]:
data['h_t_errors'].value_counts().sort_values()

h_t_errors
5       1
4      13
3      57
2     241
1     733
0    1418
Name: count, dtype: int64

In [56]:
data[data['h_t_errors'] == data['h_t_errors'].max()]

attendance             away_team  a_t_errors  a_t_hits  a_t_runs  \
1178     22581.0  Arizona Diamondbacks           1        11         8   

            date field_type   game_type          home_team  h_t_errors  \
1178  2016-07-27   on grass  Night Game  Milwaukee Brewers           5   

      h_t_hits  h_t_runs       start_time        venue day_of_week  temp_C  \
1178         5         1  7:10 p.m. Local  Miller Park   Wednesday    25.0   

      wind_speed      wind_direction     sky  total_runs  game_hours_dec  \
1178         9.0   in from Leftfield  Cloudy           9        2.933333   

              season  h_t_win  h_t_loss h_t_outcome  
1178  regular season        0         1        Loss

# 12. В какой игре было максимальное количество хитов?

Здесь я решил вынести в отдельную колонку количество хитов, а затем просто достать строку с максимальным количеством хитов. Максимальное количество хитов 37

In [59]:
data['total_hits'] = data['h_t_hits'] + data['a_t_hits']
data[data['total_hits'] == data['total_hits'].max()]

attendance      away_team  a_t_errors  a_t_hits  a_t_runs        date  \
1413     36253.0  Texas Rangers           0        16         5  2016-07-04   

     field_type game_type       home_team  h_t_errors  h_t_hits  h_t_runs  \
1413   on grass  Day Game  Boston Red Sox           2        21        12   

           start_time        venue day_of_week  temp_C  wind_speed  \
1413  1:37 p.m. Local  Fenway Park      Monday    28.9        11.0   

           wind_direction     sky  total_runs  game_hours_dec          season  \
1413   out to Centerfield  Cloudy          17        3.666667  regular season   

      h_t_win  h_t_loss h_t_outcome  total_hits  
1413        1         0         Win          37

# 13. Выведите количество игр которая сыграла каждая команда в данном сезоне?

Здесь, для удобства, я решил отделить игры в сезоне 'regular season' от игр вне сезона 'post season' при помощи метода loc. Далее поскольку идёт запись игры как 1 домашняя команда и 1 выездная, значит, что бы посчитать общее количество игр каждой команды, нужно сложить 2 столбца таблицы, в одной посчитаны команды на выезде, во второй команды, играющие дома. Итоговая таблица отражает общее количество игр, сыгранных каждой командой. Для уточнения что всё сходится подсчитаем общее количество игр и разделим на 2(так как за 1 игру играют 2 команды) и сложим с общим числом игр из 'post season'. Итоговая цифра совпадает с общим количеством строк

In [62]:
data_regular_season = data.loc[data.season == 'regular season']

In [63]:
amount_of_games = data_regular_season['away_team'].value_counts().add(data_regular_season['home_team'].value_counts())
amount_of_games

Arizona Diamondbacks             162
Atlanta Braves                   161
Baltimore Orioles                162
Boston Red Sox                   162
Chicago Cubs                     162
Chicago White Sox                162
Cincinnati Reds                  162
Cleveland Indians                161
Colorado Rockies                 162
Detroit Tigers                   161
Houston Astros                   162
Kansas City Royals               162
Los Angeles Angels of Anaheim    162
Los Angeles Dodgers              162
Miami Marlins                    161
Milwaukee Brewers                162
Minnesota Twins                  162
New York Mets                    162
New York Yankees                 162
Oakland Athletics                162
Philadelphia Phillies            162
Pittsburgh Pirates               162
San Diego Padres                 162
San Francisco Giants             162
Seattle Mariners                 162
St. Louis Cardinals              162
Tampa Bay Rays                   162
T

In [64]:
data_out_season = data.loc[data.season == 'post season']
data_out_season['season'].count() + amount_of_games.sum()/2

2463.0

# 14. Какая команда выиграла наибольшое количество матчей в сезоне? (Будьте внимательны с типом матча в сезоне - season).

В данной задаче я сделал чуть сложнее, выделил 2 таблицы с победителями на выезде и победителями домашними командами. Эти таблицы пригодятся в следующих заданиях. Затем сложил две таблицы, что бы посмотреть общее количество игры каждой команды на выезде и дома и получил команду с наибольшим общим количеством побед. Решил сделать через метод pivot_table на основе таблицы где только сезонные игры. Так же в каждой из двух полученных таблиц переименовал в колонки с одинаковым названием 'win' через метод rename, что бы можно было без проблем сложить 2 колонки из разных таблиц

In [67]:
away_team_wins = pd.pivot_table(data_regular_season, index='away_team', values='h_t_loss', aggfunc='sum')
away_team_wins.rename( columns={'h_t_loss': 'win'}, inplace=True)
away_team_wins.sort_values( by= "win", ascending=False ).head(5)

win
away_team                
St. Louis Cardinals    48
Boston Red Sox         46
Chicago Cubs           46
Washington Nationals   45
Toronto Blue Jays      43

In [68]:
home_team_wins = pd.pivot_table(data_regular_season, index='home_team', values='h_t_win', aggfunc='sum')
home_team_wins.rename( columns={'h_t_win': 'win'}, inplace=True)
home_team_wins.sort_values( by= "win", ascending=False ).head(5)

win
home_team                
Chicago Cubs           57
Cleveland Indians      53
Los Angeles Dodgers    53
Texas Rangers          53
Washington Nationals   50

In [69]:
res_table = away_team_wins.add(home_team_wins)
res_table.sort_values( by= "win", ascending=False ).head(1)

win
away_team        
Chicago Cubs  103

Получаем команду победившую в сезоне наибольшее количество раз Chicago Cubs - 103 победы


# 15. Какая команда выиграла наибольшее количество домашних матчей в сезоне?

Воспользовался уже созданной таблицей с победителями домашних игр, отсортировал по убыванию количества побед и вывел первую строку таблицы.
Получаем команду, которая выйграла наибольшее количество домашних матчей Chicago Cubs - 57 побед


In [73]:
home_team_wins.sort_values( by= "win", ascending=False ).head(1)

win
home_team        
Chicago Cubs   57

# 16. Какая команда выиграла наибольшее количество гостевых матчей в сезоне?

Аналогично с предыдущим заданием, только с гостевыми командами. St. Louis Cardinals - 48 побед в гостевых матчах.


In [76]:
away_team_wins.sort_values( by= "win", ascending=False ).head(1)

win
away_team               
St. Louis Cardinals   48

# 17. Какая команда проиграла наибольшее количество матчей в сезоне?

Здесь идём от обратного, то есть команда с наименьшим количеством побед это та, которая проигрывала больше всех, с учётом, что все команды сыграли почти одинаковое количество игр (общая разница в одну игру). Команда проигравшая больше всех тачей в сезоне Minnesota Twins

In [79]:
res_table.sort_values( by= "win", ascending=False ).tail(1)

win
away_team           
Minnesota Twins   59

In [80]:
amount_of_games['Minnesota Twins'] - res_table.sort_values( by= "win", ascending=False )['win']['Minnesota Twins']

103

Команда Minnesota Twins проиграла 103 матча

# 18. Зависит ли выигрыш от количества посетителей матча? (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.cov.html)

В данном вопросе я исследовал 2 колонки: количество посетителей ('attendance') и выгранные домашней командой матчи ('h_t_win')

In [84]:
attendance_and_h_t_win = data[['attendance', 'h_t_win']]
attendance_and_h_t_win.cov()

attendance     h_t_win
attendance  9.740607e+07  224.736752
h_t_win     2.247368e+02    0.249235

In [85]:
data['attendance'].median()

30604.0

Исходя из полученной матрицы ковариации можно сделать вывод о положительной линейной зависимости, то есть диагональ соотношений больше 0, что говорит о положительной линейной зависимости, но в то же время, на фоне общего количества посетителей (медианное значение которых 30604), изменение в 224 посетителя можно считать как стремящейся к нулю, значит положительную тенденцию можно рассматривать как статистическую погрешность

# 19. Правда ли что большинство проигрышных домашних матчей приходятся на Субботу и Воскресенье?

Здесь я создал функцию, которая разбивает дни недели на рабочие и выходные, полученные данные мы передаём в новую колонку 'weekday', для проверки выведем первые 5 записей и сверим, что функция работает корректно. 

In [89]:
def sort_week_days(row):
    if row['day_of_week'].lower() in ['saturday', 'sunday',]:
        return 'weekend'
    else:
        return 'workday'

data['weekday'] = data.apply(sort_week_days, axis=1)
data.head()

attendance              away_team  a_t_errors  a_t_hits  a_t_runs  \
0     40030.0          New York Mets           1         7         3   
1     21621.0  Philadelphia Phillies           0         5         2   
2     12622.0        Minnesota Twins           0         5         2   
3     18531.0   Washington Nationals           0         8         3   
4     18572.0       Colorado Rockies           1         8         4   

         date field_type   game_type             home_team  h_t_errors  \
0  2016-04-03   on grass  Night Game    Kansas City Royals           0   
1  2016-04-06   on grass  Night Game       Cincinnati Reds           0   
2  2016-04-06   on grass  Night Game     Baltimore Orioles           0   
3  2016-04-06   on grass  Night Game        Atlanta Braves           1   
4  2016-04-06   on grass    Day Game  Arizona Diamondbacks           0   

   h_t_hits  h_t_runs        start_time                        venue  \
0         9         4   7:38 p.m. Local             Kauffman Stadium   
1         8         3   7:11 p.m. Local     Great American Ball Park   
2         9         4   7:07 p.m. Local  Oriole Park at Camden Yards   
3         8         1   7:10 p.m. Local                 Turner Field   
4         8         3  12:40 p.m. Local                  Chase Field   

  day_of_week  temp_C  wind_speed         wind_direction       sky  \
0      Sunday    23.3        14.0     from Right to Left     Sunny   
1   Wednesday    12.8        24.0     from Right to Left  Overcast   
2   Wednesday     8.9         7.0       out to Leftfield   Unknown   
3   Wednesday    18.3        10.0     from Right to Left    Cloudy   
4   Wednesday    25.0         0.0   in unknown direction   In Dome   

   total_runs  game_hours_dec          season  h_t_win  h_t_loss h_t_outcome  \
0           7        3.216667  regular season        1         0         Win   
1           5        2.383333  regular season        1         0         Win   
2           6        3.183333  regular season        1         0         Win   
3           4        2.883333  regular season        0         1        Loss   
4           7        2.650000  regular season        0         1        Loss   

   total_hits  weekday  
0          16  weekend  
1          13  workday  
2          14  workday  
3          16  workday  
4          16  workday

После через pivot_table выделяем созданную колонку, передаём как значение количество проигрышей домашних команд и количество побед домашних команд. Объеденяем полученные таблицы в одну при помощи метода merge и передаём объеденяющий ключ on=["weekday"], считаем общее количество игр и процент побед от общего количества.

In [91]:
weekday_comparison = pd.pivot_table( data, index = 'weekday', values = 'h_t_loss', aggfunc='sum')
weekday_comparison_wins = pd.pivot_table( data, index = 'weekday', values = 'h_t_win', aggfunc='sum')

In [92]:
result_table = weekday_comparison.merge(weekday_comparison_wins, on=["weekday"])
result_table['amount_of_games'] = result_table['h_t_loss'] + result_table['h_t_win']
result_table['%_loss'] = result_table['h_t_loss'] * 100 / result_table['amount_of_games']
result_table['%_wins'] = result_table['h_t_win'] * 100 / result_table['amount_of_games']
result_table

h_t_loss  h_t_win  amount_of_games     %_loss     %_wins
weekday                                                          
weekend       365      423              788  46.319797  53.680203
workday       793      881             1674  47.371565  52.628435

Из приведённой таблицы можно сделать вывод, что процент побед в целом доманшей команды выше, чем у гостевой команды. Но вопрос был про проигрыш домашней команды на выходные дни. Поскольку мы видим, что на выходные процент побед немного увеличивается, а процент проигрышей уменьшается, следовательно ответ НЕТ. 

# 20. Правда ли что наибольшее количество ранов происходит в холодную погоду? (Холодной погодой считается погода ниже 0 градусов)

Создаём маску для отбора игр, при температуре ниже нуля и применяем его через метод loc

In [96]:
mask = (data.temp_C < 0)
data.loc[ mask ]

attendance         away_team  a_t_errors  a_t_hits  a_t_runs  \
2409     32419.0  New York Yankees           1        13         8   

            date field_type game_type       home_team  h_t_errors  h_t_hits  \
2409  2016-04-09   on grass  Day Game  Detroit Tigers           1         7   

      h_t_runs       start_time          venue day_of_week  temp_C  \
2409         4  1:10 p.m. Local  Comerica Park    Saturday    -0.6   

      wind_speed       wind_direction     sky  total_runs  game_hours_dec  \
2409        18.0   from Left to Right  Cloudy          12        3.333333   

              season  h_t_win  h_t_loss h_t_outcome  total_hits  weekday  
2409  regular season        0         1        Loss          20  weekend

In [97]:
data.loc[ mask ]['total_runs']

2409    12
Name: total_runs, dtype: int64

In [98]:
data['total_runs'].median()

8.0

Согласно пересчёту Фаренгейтов в градусы Цельсия, только одна игра проходила при температуре ниже 0 градусов, за которую было сделано всего 12 ранов. Это больше чем медианное значение, но мы не можем утверждать что количество ранов больше в холодную погоду основываясь только на одной игре, в то время как игр, проведённых при температуре больше нуля - 2462. В данном случае можно сказать только о недостатке данных, для проведения такого рода исследования.